docs/user-defined-type-functions.md

```diff
@@ -0,0 +1,235 @@
+| `isbooleansingleton()` | `boolean` | returns true if self is a runtime representation of a boolean singleton |
+| `isunion()` | `boolean` | returns true if self is a runtime representation of the union type |
+| `isintersection()` | `boolean` | returns true if self is a runtime representation of the intersection type |
+| `istable()` | `boolean` | returns true if self is a runtime representation of a `table` type |
+| `ismetatable()` | `boolean` | returns true if self is a runtime representation of `Metatable` |
+| `isfunction()` | `boolean` | returns true if self is a runtime representation of a `function` type |
+| `isclass()` | `boolean` | returns true if self is a runtime representation of a `class` type |
+| `issubtypeof(arg: typelib)` | `boolean` | returns true if self is a subtype or equal to arg in the type hierarchy |
+| `conformsto(arg: typelib)` | `boolean` | returns true if self is equal to arg in the type hierarchy |
+
+#### Negation
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `gettype()` | `typelib` | returns the runtime representation of the type being negated |
+
+#### String
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getmetatable()` | `typelib` | returns the runtime representation of `Metatable` |
+
+#### StringSingleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `string` | returns a string singleton |
+
+#### BooleanSingleton
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `getvalue()` | `boolean` | returns either `true` or `false` |
+
+#### Table
+
+| Function Declaration | Return Type | Description |
+| ------------- | ------------- | ------------- |
+| `addprops(key: typelib, value: typelib)` | `nil` | adds a key, value pair to self's table properties; if the same key exists already, overrides the value |
+| `delprops(key: typelib)` | `nil` | removes the key from self's table properties along with the value associated with it; if the key doesn't exist, nothing happens |
+| `getprops()` | `{typelib: typelib}` | returns a table of self's table properties (e.g. `{["age"] = 20}` will return `{typelib:getstringsingleton("age") = typelib:getnumber()}`) |
+| `settableindexer(indexty: typelib, indexresultty: typelib)` | `nil` | sets self's index type to the first argument and index result type to the second |
+| `getindextype()` | `typelib` | returns self's index type |
+| `getindexresulttype()` | `typelib` | returns self's index result type |
+| `setsealed(arg: boolean)` | `nil` | sets the table's sealed status. If true, the table becomes a sealed table. Default initialization of tables is unsealed tables (aka setsealed(false)) |
+| `issealed()` | `boolean` | returns the seal status of self |
```

Comment:
I don't think either of these should be exposed to users at all.

Reply:
Yeah. Absolutely an implementation detail.